src/kernel/klib.c

```diff
@@ -257,14 +319,34 @@ void init_klib(kernel_heaps kh, void *fs, tuple config_root, tuple klib_md)
                klib_heap_start, klib_heap_size);
     klib_heap = create_id_heap(h, h, klib_heap_start, klib_heap_size, PAGESIZE, false);
     assert(klib_heap != INVALID_ADDRESS);
+    tuple c = get_tuple(klib_md, sym(children));
+    if (!c)
+        goto done;
+    klib_root = get_tuple(c, sym(klib));
+    if (!klib_root)
+        goto done;
     if (get(config_root, sym(klib_test))) {
         klib_debug("   loading klib test\n");
-        load_klib("/klib/test", closure(h, klib_test_loaded));
+        load_klib("test/test", closure(h, klib_test_loaded), 0);
+    }
+    c = get_tuple(klib_root, sym(children));
+    if (c) {
+        u64 klib_count = 0;
+        iterate(c, stack_closure(count_klib_each, &klib_count));
+        if (klib_count == 0)
+            goto done;
+        queue retry_klibs = allocate_queue(h, klib_count);
+        assert(retry_klibs != INVALID_ADDRESS);
+        merge m = allocate_merge(h, complete);
+        status_handler sh;
+        do {
+            sh = apply_merge(m);
```

Comment:
The status_handler returned by apply_merge() is always the same, and since we already checked for (klib_count == 0), we are sure that the loop is executed at least once and so the sh variable is initialized. Then, `sh` is passed to the autoload_klib_complete and autoload_klib_each closures, and from there it will be applied at some point, thus closing the merge when appropriate.

Reply:
Sorry for the delayed response. The only issue I had with the merge use there is that, as I understood it - or thought I did, anyway - the merge datatype is meant to deal out status_handlers for some number of asynchronous operations, and that the fact that the status handlers are all one in the same is just an implementation detail at the present moment. I know this isn't documented, and I guess it seems kind of nit-picky to request that they be treated as separate, but I just think that would be a safer option - in the case there might ever be a merge implementation (or mode) where different handlers are dealt out, however remote it may seem. For example, there could one day be a debug version of a merge which uses individual closures to associate statuses with the operation (by way of sequence maybe, and also apply_merge caller), or there could be an alternate completion type for a merge which reports such associations. Does this make any sense? Do you think it would be a hassle to revise this to operate as if the status_handlers are unique?

Reply:
OK, I agree it's safer not to assume anything about the status handlers returned by apply_merge(). I modified the code so that autoload_klib_each() calls apply_merge() and passes its return value to autoload_klib_complete (which is now allocated separately for each klib). The separate counting of klibs to be auto-loaded (which was there to avoid leaking the retry_klibs queue when there are no klibs to be loaded) has been removed, since the cleanup steps are now done by klibs_complete(), which is called as a completion of the merge.